pkg/registry/generic/etcd/etcd.go

```diff
@@ -407,30 +403,30 @@ func (e *Etcd) finalizeDelete(obj runtime.Object, runHooks bool) (runtime.Object
 	return &api.Status{Status: api.StatusSuccess}, nil
 }
 
-// WatchPredicate starts a watch for the items that m matches.
+// Watch makes a matcher for the given label and field, and calls
+// WatchPredicate. If possible, you should customize PredicateFunc to produre a
+// matcher that matches by key.
 func (e *Etcd) Watch(ctx api.Context, label labels.Selector, field fields.Selector, resourceVersion string) (watch.Interface, error) {
-	if value, found := field.RequiresExactMatch(e.WatchSingleFieldName); found && len(e.WatchSingleFieldName) > 0 {
-		key, err := e.KeyFunc(ctx, value)
-		if err != nil {
-			return nil, err
-		}
-		return e.watchPredicate(key, e.PredicateFunc(label, field), resourceVersion)
-	}
-	return e.watchPredicate(e.KeyRootFunc(ctx), e.PredicateFunc(label, field), resourceVersion)
+	return e.WatchPredicate(ctx, e.PredicateFunc(label, field), resourceVersion)
 }
 
 // WatchPredicate starts a watch for the items that m matches.
-// TODO: Detect if m references a single object instead of a list.
 func (e *Etcd) WatchPredicate(ctx api.Context, m generic.Matcher, resourceVersion string) (watch.Interface, error) {
-	return e.watchPredicate(e.KeyRootFunc(ctx), m, resourceVersion)
-}
-
-func (e *Etcd) watchPredicate(key string, m generic.Matcher, resourceVersion string) (watch.Interface, error) {
 	version, err := tools.ParseWatchResourceVersion(resourceVersion, e.EndpointName)
 	if err != nil {
 		return nil, err
 	}
-	return e.Helper.WatchList(key, version, func(obj runtime.Object) bool {
+
+	if k, ok := m.MatchesSingle(); ok {
+		key, err := e.KeyFunc(ctx, k)
+		if err != nil {
+			return nil, err
+		}
+		// This matches a single object-- no need to call the predicate all the time.
```

Comment:
I'm wondering if that's always true. It's technically possible that you specify to match a single object that additionally matches some labels. Don't we want to support that case too? (i.e. don't we want to call this predicate anyway even if we are watching a single object?)